compiler/src/iree/compiler/Dialect/Stream/Analysis/Partitioning/ReferencePartitioning.cpp

```diff
@@ -52,24 +60,76 @@ partitionStreamableOpsReference(IREE::Stream::PartitioningConfigAttr config,
     SetVector<Operation *> ops;
     // Ops that were cloned and are known not to have their values escape.
     DenseSet<Operation *> clonedOps;
-    void insert(Operation *op) {
+    // Which partitions transitively depend on this partition.
+    llvm::BitVector hazards;
+    void insert(Operation *op, OpInfo &opInfo) {
       if (auto affinityOp = dyn_cast<IREE::Stream::AffinityOpInterface>(op)) {
         affinity = affinity ? affinity.joinAND(affinityOp.getAffinityAttr())
                             : affinityOp.getAffinityAttr();
       }
+      opInfo.membership.set(ordinal);
+      if (opInfo.hazards.size() > ordinal)
+        opInfo.hazards.reset(ordinal);
       ops.insert(op);
+      hazards |= opInfo.hazards;
     }
   };
   SmallVector<std::unique_ptr<PartitionBuilder>> builders;
   llvm::BitVector usableBuilders;
 
-  struct OpInfo {
-    // Which partitions the op is contained within.
-    llvm::BitVector membership;
-    // Which partitions transitively depend on this operation.
-    llvm::BitVector hazards;
+  auto willCreateCircularDependencyBetweenPartitions =
+      [&](unsigned sourceOrdinal, unsigned targetOrdinal) -> bool {
+    // Returns:
+    // If we are to make partition with ordinal targetOrdinal to
+    // depend on partition with ordinal sourceOrdinal,
+    // will this create a circular dependency.
+    if (sourceOrdinal == targetOrdinal)
+      return false;
+    return builders[sourceOrdinal]->hazards.size() > targetOrdinal &&
+           builders[sourceOrdinal]->hazards[targetOrdinal];
+  };
+
+  auto canAddOpToPartition = [&](Operation &op, OpInfo &opInfo,
+                                 unsigned partitionOrdinal) {
+    auto streamableOp = dyn_cast<IREE::Stream::StreamableOpInterface>(op);
+    if (!streamableOp)
+      return false;
+    IREE::Stream::AffinityAttr affinityAttr;
+    if (auto affinityOp = dyn_cast<IREE::Stream::AffinityOpInterface>(op))
+      affinityAttr = affinityOp.getAffinityAttr();
+    if (!IREE::Stream::AffinityAttr::canExecuteTogether(
+            affinityAttr, builders[partitionOrdinal]->affinity))
+      return false;
+
+    bool preferCloneToConsumers = streamableOp.preferCloneToConsumers();
+    llvm::BitVector *opHazards = nullptr;
+    llvm::BitVector opHazardsInCandidatePartition;
+    if (preferCloneToConsumers) {
+      // If we are cloning we care only about users that are a part of the
+      // candidate partition.
+      opHazards = &opHazardsInCandidatePartition;
+      for (auto user : op.getUsers()) {
+        if (builders[partitionOrdinal]->ops.contains(user))
+          opHazardsInCandidatePartition |= opInfos[user].hazards;
+      }
```

Comment:
Here actually, we would need to walk further down the users if a user is also cloned into the partition.